include/seastar/core/semaphore.hh

```diff
@@ -475,8 +475,11 @@ public:
     semaphore_units(semaphore_units&& o) noexcept : _sem(o._sem), _n(std::exchange(o._n, 0)) {
     }
     semaphore_units& operator=(semaphore_units&& o) noexcept {
-        _sem = o._sem;
-        _n = std::exchange(o._n, 0);
+        if (this != &o) {
+            return_all();
+            _sem = o._sem;
+            _n = std::exchange(o._n, 0);
```

Comment:
nit:
i understand the fix and why it is correct. 
what seems weird is that the proper pattern should be 
```
if (this!= &o){
  std::swap(_sem, o._sem);
  std::swap(_n, o._n);
}
```
and then the destructor of `o` would take care of correctness. i wonder if I'm missing something